File: app/src/main/java/jp/juggler/subwaytooter/action/Action_ListMember.java
package jp.juggler.subwaytooter.action;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.regex.Pattern;

import jp.juggler.subwaytooter.ActMain;
import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.Column;
import jp.juggler.subwaytooter.R;
import jp.juggler.subwaytooter.api.TootApiClient;
import jp.juggler.subwaytooter.api.TootApiResult;
import jp.juggler.subwaytooter.api.TootTask;
import jp.juggler.subwaytooter.api.TootTaskRunner;
import jp.juggler.subwaytooter.api.entity.TootAccount;
import jp.juggler.subwaytooter.api.entity.TootRelationShip;
import jp.juggler.subwaytooter.dialog.DlgConfirm;
import jp.juggler.subwaytooter.table.SavedAccount;
import jp.juggler.subwaytooter.util.Utils;
import okhttp3.Request;
import okhttp3.RequestBody;

public class Action_ListMember {
	
	public interface Callback {
		void onListMemberUpdated( boolean willRegistered, boolean bSuccess );
	}
	
	private static final Pattern reFollowError = Pattern.compile( "follow", Pattern.CASE_INSENSITIVE );
	
	public static void add(
		@NonNull final ActMain activity
		, @NonNull final SavedAccount access_info
		, final long list_id
		, @NonNull final TootAccount local_who
		, final boolean bFollow
		, @Nullable final Callback callback
	){
		new TootTaskRunner( activity,  true ).run( access_info, new TootTask() {
			@Override public TootApiResult background( @NonNull TootApiClient client ){
				
				if( access_info.isMe( local_who ) ){
					return new TootApiResult( activity.getString( R.string.it_is_you ) );
				}
				
				TootApiResult result;
				
				if( bFollow ){
					TootRelationShip relation;
					if( access_info.isLocalUser( local_who ) ){
						Request.Builder request_builder = new Request.Builder().post(
							RequestBody.create(
								TootApiClient.MEDIA_TYPE_FORM_URL_ENCODED
								, "" // 空データ
							) );
						
						result = client.request( "/api/v1/accounts/" + local_who.id + "/follow", request_builder );
					}else{
						// リモートフォローする
						Request.Builder request_builder = new Request.Builder().post(
							RequestBody.create(
								TootApiClient.MEDIA_TYPE_FORM_URL_ENCODED
								, "uri=" + Uri.encode( local_who.acct )
							) );
						
						result = client.request( "/api/v1/follows", request_builder );
						if( result == null || result.object == null ) return result;
						
						TootAccount a = TootAccount.parse( activity, access_info, result.object );
						if( a == null ){
							return new TootApiResult( "parse error." );
						}
						
						// リモートフォローの後にリレーションシップを取得しなおす
						result = client.request( "/api/v1/accounts/relationships?id[]=" + a.id );
					}
					
					if( result == null || result.array == null ){
						return result;
					}
					
					TootRelationShip.List relation_list = TootRelationShip.parseList( result.array );
					relation = relation_list.isEmpty() ? null : relation_list.get( 0 );
					
					if( relation == null ){
						return new TootApiResult( "parse error." );
					}
					ActionUtils.saveUserRelation( access_info, relation );
					
					if( ! relation.following ){
						if( relation.requested ){
							return new TootApiResult( activity.getString( R.string.cant_add_list_follow_requesting ) );
						}else{
							// リモートフォローの場合、正常ケースでもここを通る場合がある
							// 何もしてはいけない…
						}
					}
				}
				
				// リストメンバー追加
				
				JSONObject content = new JSONObject();
				try{
					JSONArray account_ids = new JSONArray();
					account_ids.put( Long.toString( local_who.id ) );
					content.put( "account_ids", account_ids );
				}catch( Throwable ex ){
					return new TootApiResult( Utils.formatError( ex, "can't encoding json parameter." ) );
				}
				
				Request.Builder request_builder = new Request.Builder().post(
					RequestBody.create(
						TootApiClient.MEDIA_TYPE_JSON
						, content.toString()
					) );
				
				return client.request( "/api/v1/lists/" + list_id + "/accounts", request_builder );
				
			}
			
			@Override public void handleResult( @Nullable TootApiResult result ){
				boolean bSuccess = false;
				
				try{
					//noinspection StatementWithEmptyBody
					if( result == null ) return; // cancelled.
					
					if( result.object != null ){
						for( Column column : App1.app_state.column_list ){
							// リストメンバー追加イベントをカラムに伝達
							column.onListMemberUpdated( access_info, list_id, local_who, true );
						}
						// フォロー状態の更新を表示に反映させる
						if( bFollow ) activity.showColumnMatchAccount( access_info );
						
						Utils.showToast( activity, false, R.string.list_member_added );
						
						bSuccess = true;
						
					}else{
						
						if( result.response != null
							&& result.response.code() == 422
							&& result.error != null && reFollowError.matcher( result.error ).find()
							){
							
							if( ! bFollow ){
								DlgConfirm.openSimple(
									activity
									, activity.getString( R.string.list_retry_with_follow, access_info.getFullAcct( local_who ) )
									, new Runnable() {
										@Override public void run(){
											Action_ListMember.add( activity, access_info, list_id, local_who, true, callback );
										}
									}
								);
							}else{
								new android.app.AlertDialog.Builder( activity )
									.setCancelable( true )
									.setMessage( R.string.cant_add_list_follow_requesting )
									.setNeutralButton( R.string.close, null )
									.show();
							}
							return;
						}
						
						Utils.showToast( activity, true, result.error );
						
					}
				}finally{
					if( callback != null ) callback.onListMemberUpdated( true, bSuccess );
				}
				
			}
		} );
	}
	
	public static void delete(
		@NonNull final ActMain activity
		, @NonNull final SavedAccount access_info
		, final long list_id
		, @NonNull final TootAccount local_who
		, @Nullable final Callback callback
	){
		new TootTaskRunner( activity,  true ) .run( access_info, new TootTask() {
			@Override public TootApiResult background( @NonNull TootApiClient client ){
				return client.request(
					"/api/v1/lists/" + list_id + "/accounts?account_ids[]=" + local_who.id
					, new Request.Builder().delete()
				);
			}
			
			@Override public void handleResult( @Nullable TootApiResult result ){
				boolean bSuccess = false;
				
				try{
					
					if( result == null ) return; // cancelled.
					
					if( result.object != null ){
						
						for( Column column : App1.app_state.column_list ){
							column.onListMemberUpdated( access_info, list_id, local_who, false );
						}
						
						Utils.showToast( activity, false, R.string.delete_succeeded );
						
						bSuccess = true;
						
					}else{
						Utils.showToast( activity, false, result.error );
					}
				}finally{
					if( callback != null ) callback.onListMemberUpdated( false, bSuccess );
				}
				
			}
		} );
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/action/Action_ListMember.kt
package jp.juggler.subwaytooter.action

import android.app.AlertDialog
import jp.juggler.subwaytooter.ActMain
import jp.juggler.subwaytooter.R
import jp.juggler.subwaytooter.actmain.showColumnMatchAccount
import jp.juggler.subwaytooter.api.TootApiResult
import jp.juggler.subwaytooter.api.TootParser
import jp.juggler.subwaytooter.api.entity.*
import jp.juggler.subwaytooter.api.runApiTask
import jp.juggler.subwaytooter.api.syncAccountByAcct
import jp.juggler.subwaytooter.column.onListMemberUpdated
import jp.juggler.subwaytooter.dialog.DlgConfirm.confirm
import jp.juggler.subwaytooter.table.SavedAccount
import jp.juggler.subwaytooter.table.daoUserRelation
import jp.juggler.util.coroutine.launchMain
import jp.juggler.util.data.asciiPattern
import jp.juggler.util.data.buildJsonArray
import jp.juggler.util.data.buildJsonObject
import jp.juggler.util.data.notBlank
import jp.juggler.util.data.notEmpty
import jp.juggler.util.log.LogCategory
import jp.juggler.util.log.showToast
import jp.juggler.util.network.*
import kotlinx.coroutines.CancellationException
import okhttp3.Request
import java.util.regex.Pattern

private val log = LogCategory("Action_ListMember")
private val reFollowError422 = "follow".asciiPattern(Pattern.CASE_INSENSITIVE)
private val reFollowError404 = "Record not found".asciiPattern(Pattern.CASE_INSENSITIVE)

class MemberNotFollowedException : IllegalStateException()

/**
 * リストに追加する。失敗したら例外を投げる。
 */
suspend fun ActMain.listMemberAdd(
    accessInfo: SavedAccount,
    listId: EntityId,
    localWho: TootAccount,
) {
    var i = 0
    while(true){
        ++i
        val result = runApiTask(accessInfo) { client ->
            val parser = TootParser(this, accessInfo)

            var userId = localWho.id

            if (accessInfo.isMisskey) {
                // misskeyのリストはフォロー無関係
                client.request(
                    "/api/users/lists/push",
                    accessInfo.putMisskeyApiToken().apply {
                        put("listId", listId)
                        put("userId", localWho.id)
                    }.toPostRequestBuilder()
                )
                // 204 no content
            } else {
                val isMe = accessInfo.isMe(localWho)
                if (isMe) {
                    val (ti, ri) = TootInstance.get(client)
                    ti ?: return@runApiTask ri
                    if (!ti.versionGE(TootInstance.VERSION_3_1_0_rc1)) {
                        return@runApiTask TootApiResult(getString(R.string.it_is_you))
                    }
                } else if (i ==1) {
                    // リモートユーザの解決
                    if (!accessInfo.isLocalUser(localWho)) {
                        val (r2, ar) = client.syncAccountByAcct(accessInfo, localWho.acct)
                        val user = ar?.get() ?: return@runApiTask r2
                        userId = user.id
                    }

                    val result = client.request(
                        "/api/v1/accounts/$userId/follow",
                        "".toFormRequestBody().toPost()
                    ) ?: return@runApiTask null

                    val relation = daoUserRelation.saveUserRelation(
                        accessInfo,
                        parseItem(result.jsonObject) { TootRelationShip(parser, it) }
                    ) ?: return@runApiTask TootApiResult("parse error.")

                    if (!relation.following) {
                        if (relation.requested) {
                            return@runApiTask TootApiResult(getString(R.string.cant_add_list_follow_requesting))
                        } else {
                            // リモートフォローの場合、正常ケースでもここを通る場合がある
                            // 何もしてはいけない…
                        }
                    }
                }

                // リストメンバー追加
                client.request(
                    "/api/v1/lists/$listId/accounts",
                    buildJsonObject {
                        put(
                            "account_ids",
                            buildJsonArray {
                                add(userId.toString())
                            }
                        )
                    }.toPostRequestBuilder()
                )
            }
        }

        // cancelled.
        result ?: throw CancellationException()

        // 成功
        if( result.jsonObject != null) break

        val response = result.response
        val errorMessage = result.error
        val isNotFollowed = response != null && errorMessage != null && when {
            response.code == 422 && reFollowError422.matcher(errorMessage).find() -> true
            response.code == 404 && reFollowError404.matcher(errorMessage).find() -> true
            else -> false
        }
        when {
            isNotFollowed -> {
                when {
                    // 初回ならフォロー操作を行うかユーザに尋ねてリトライ
                    i==1 -> {
                        confirm(
                            R.string.list_retry_with_follow,
                            accessInfo.getFullAcct(localWho)
                        )
                        continue
                    }
                    // 2回目以降は諦める
                    else ->{
                        // フォロー操作を行ってもダメなら失敗とみなす
                        // フォロー要求が承認されるまでラグがあるかもしれないので、ユーザに伝える
                        AlertDialog.Builder(this@listMemberAdd)
                            .setCancelable(true)
                            .setMessage(R.string.cant_add_list_follow_requesting)
                            .setNeutralButton(R.string.close, null)
                            .show()
                        throw MemberNotFollowedException()
                    }
                }
            }
            // その他のエラー
            else -> error(errorMessage.notBlank() ?: "?")
        }
    }
    // 成功後
    for (column in appState.columnList) {
        // リストメンバー追加イベントをカラムに伝達
        column.onListMemberUpdated(accessInfo, listId, localWho, true)
    }
    // フォロー状態の更新を表示に反映させる
    if (i > 0) showColumnMatchAccount(accessInfo)
    showToast(false, R.string.list_member_added)
}

suspend fun ActMain.listMemberDelete(
    accessInfo: SavedAccount,
    listId: EntityId,
    localWho: TootAccount,
) {
    val result = runApiTask(accessInfo) { client ->
        if (accessInfo.isMisskey) {
            client.request(
                "/api/users/lists/pull",
                accessInfo.putMisskeyApiToken().apply {
                    put("listId", listId.toString())
                    put("userId", localWho.id.toString())
                }.toPostRequestBuilder()
            )
        } else {
            client.request(
                "/api/v1/lists/$listId/accounts?account_ids[]=${localWho.id}",
                Request.Builder().delete()
            )
        }
    }
    when{
        // キャンセル
        result == null -> throw CancellationException()

        // 失敗
        result.jsonObject == null -> error(result.error.notBlank() ?: "?")

        // 成功
        else ->{
            for (column in appState.columnList) {
                column.onListMemberUpdated(accessInfo, listId, localWho, false)
            }
            showToast(false, R.string.delete_succeeded)
        }
    }
}
